mobilecoind-dev-faucet/src/worker.rs

```diff
@@ -0,0 +1,632 @@
+    ///   a queue falls below this number the worker attempts a split Tx.
+    /// * worker_poll_period: A lower bound on how often the worker should poll
+    /// * logger
+    ///
+    /// Returns the Worker handle object, which contains the thread handle, and
+    /// receives the output of the worker thread.
+    pub fn new(
+        client: MobilecoindApiClient,
+        monitor_id: Vec<u8>,
+        public_address: PublicAddress,
+        minimum_fees: HashMap<TokenId, u64>,
+        target_amounts: HashMap<TokenId, u64>,
+        target_queue_depth: usize,
+        worker_poll_period: Duration,
+        logger: &Logger,
+    ) -> Worker {
+        let mut worker_token_states = Vec::<WorkerTokenState>::default();
+        let mut receivers = HashMap::<TokenId, TokenStateReceiver>::default();
+
+        for (token_id, value) in target_amounts.iter() {
+            let minimum_fee_value = minimum_fees
+                .get(token_id)
+                .unwrap_or_else(|| panic!("Missing minimum fee for {}", token_id));
+            let (state, receiver) = WorkerTokenState::new(*token_id, *minimum_fee_value, *value);
+            worker_token_states.push(state);
+            receivers.insert(*token_id, receiver);
+        }
+
+        let stop_requested = Arc::new(AtomicBool::default());
+        let thread_stop_requested = stop_requested.clone();
+
+        let logger = logger.new(o!("thread" => "worker"));
+        let join_handle = Some(std::thread::spawn(move || {
+            Self::worker_thread_entry_point(
+                worker_token_states,
+                thread_stop_requested,
+                client,
+                monitor_id,
+                public_address,
+                target_queue_depth,
+                worker_poll_period,
+                logger,
+            )
+        }));
+
+        Worker {
+            receivers,
+            join_handle,
+            stop_requested,
+        }
+    }
+
+    /// The entrypoint for the worker thread.
+    /// First, wait for account to sync in mobilecoind.
+    /// Then enter a loop where we check stop_requested, poll each token id for
+    /// activity, and sleep for a bit.
+    fn worker_thread_entry_point(
+        mut worker_token_states: Vec<WorkerTokenState>,
+        stop_requested: Arc<AtomicBool>,
+        client: MobilecoindApiClient,
+        monitor_id: Vec<u8>,
+        public_address: PublicAddress,
+        target_queue_depth: usize,
+        worker_poll_period: Duration,
+        logger: Logger,
+    ) {
+        // First wait for account to sync
+        // Get the "initial" ledger block count
+        let block_count = loop {
+            match client.get_ledger_info(&Default::default()) {
+                Ok(resp) => break resp.block_count,
+                Err(err) => {
+                    log::error!(logger, "Could not get ledger info: {:?}", err);
+                }
+            }
+            std::thread::sleep(worker_poll_period);
+        };
```

Comment:
I guess that it will log errors at a rate of `worker_poll_period`, unless / until mobilecoind recovers

That seems like desirable behavior, what would you rather see?

Reply:
We could backoff on the timeout so that we don't log at faster than 1 Hz or something?